src/TrieSet.mo

```diff
@@ -38,19 +74,34 @@ module {
 
   /// Test if two sets are equal.
   public func equal<T>(s1 : Set<T>, s2 : Set<T>, eq : (T, T) -> Bool) : Bool {
-    // XXX: Todo: use a smarter check
-    func unitEqual(_ : (), _ : ()) : Bool { true };
-    Trie.equalStructure<T, ()>(s1, s2, eq, unitEqual)
+    if (Trie.size(s1) != Trie.size(s2)) return false;
+    for (k in keys(s1)) {
+      if (Trie.find<T,()>(s2, k, eq) == null) {
+        return false;
+      }
+    };
+    return true;
   };
 
   /// The number of set elements, set's cardinality.
   public func size<T>(s : Set<T>) : Nat {
-    Trie.foldUp<T, (), Nat>(
-      s,
-      func(n : Nat, m : Nat) : Nat { n + m },
-      func(_ : T, _ : ()) : Nat { 1 },
-      0
-    )
+    Trie.size(s);
+  };
+
+  /// Test if `s` is the empty set.
+  public func isEmpty<T>(s : Set<T>) : Bool {
+    Trie.size(s) == 0;
+  };
+
+  /// Test if `s1` is a subset of `s2`.
+  public func isSubset<T>(s1 : Set<T>, s2 : Set<T>, eq : (T, T) -> Bool) : Bool {
+    if (Trie.size(s1) > Trie.size(s2)) return false;
+    for (k in keys(s1)) {
+      if (Trie.find<T,()>(s2, k, eq) == null) {
+        return false;
+      }
+    };
+    return true;
```

Comment:
Similar comment to the `equal` optimization.

I think you can flatten both sets to an array (using the same ordering) and then this problem reduces to the sorted arrays subsequence problem which can be solved in `O(size1 + size2)` time.

Reply:
Same comment as my comment above: The structure of the `Trie` can be exploited here, though I'm not sure how much that would help.